internal/p2p/client/client.go

```diff
@@ -0,0 +1,311 @@
+		// GetSyncStatus requests a block synchronization status from all connected peers
+		GetSyncStatus(ctx context.Context) error
+	}
+	// Client is a stateful implementation of a client, which means that the client stores a request ID
+	// in order to be able to resolve the response once it is received from the peer
+	Client struct {
+		chanStore      *chanStore
+		clock          clockwork.Clock
+		logger         log.Logger
+		pending        sync.Map
+		reqTimeout     time.Duration
+		chanIDResolver func(msg proto.Message) p2p.ChannelID
+	}
+	// OptionFunc is a client optional function, it is used to override the default parameters in a Client
+	OptionFunc func(c *Client)
+	result     struct {
+		Value any
+		Err   error
+	}
+)
+
+// WithLogger is an optional function to set logger to Client
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(c *Client) {
+		c.logger = logger
+	}
+}
+
+// WithClock is an optional function to set clock to Client
+func WithClock(clock clockwork.Clock) OptionFunc {
+	return func(c *Client) {
+		c.clock = clock
+	}
+}
+
+// WithChanIDResolver is an option function to set channel ID resolver function
+func WithChanIDResolver(resolver func(msg proto.Message) p2p.ChannelID) OptionFunc {
+	return func(c *Client) {
+		c.chanIDResolver = resolver
+	}
+}
+
+// New creates and returns Client with optional functions
+func New(descriptors map[p2p.ChannelID]*p2p.ChannelDescriptor, creator p2p.ChannelCreator, opts ...OptionFunc) *Client {
+	client := &Client{
+		chanStore:      newChanStore(descriptors, creator),
+		clock:          clockwork.NewRealClock(),
+		logger:         log.NewNopLogger(),
+		reqTimeout:     peerTimeout,
+		chanIDResolver: p2p.ResolveChannelID,
+	}
+	for _, opt := range opts {
+		opt(client)
+	}
+	return client
+}
+
+// GetBlock requests a block from a peer and returns promise.Promise which resolve the result
+// if response received in time otherwise reject
+func (c *Client) GetBlock(ctx context.Context, height int64, peerID types.NodeID) (*promise.Promise[*bcproto.BlockResponse], error) {
+	reqID := uuid.NewString()
+	err := c.Send(ctx, p2p.Envelope{
+		ChannelID:  p2p.BlockSyncChannel,
+		Attributes: map[string]string{RequestIDAttribute: reqID},
+		To:         peerID,
+		Message:    &bcproto.BlockRequest{Height: height},
+	})
+	if err != nil {
+		errSendError := c.Send(ctx, p2p.PeerError{
+			NodeID: peerID,
+			Err:    err,
+		})
+		if errSendError != nil {
+			return nil, multierror.Append(err, errSendError)
+		}
+	}
+	respCh := c.addPending(reqID)
+	return newPromise[*bcproto.BlockResponse](ctx, peerID, reqID, respCh, c), nil
+}
+
+// GetSyncStatus requests a block synchronization status from all connected peers
+// Since this is broadcast request, we can't use promise to process a response
+// instead, we should be able to process the response as a normal message in the handler
+func (c *Client) GetSyncStatus(ctx context.Context) error {
```

Comment:
My idea was to create p2p client that provide the methods to interact with other nodes through p2p.
The client should implement (in the future) all available request (and response) operations.
it would be easier to maintain this code in one place than to create a copy of the client in each module separately